src/System.Net.Http/src/System/Net/Http/Managed/AuthenticateAndRedirectHandler.cs

```diff
@@ -0,0 +1,223 @@
+            {
+                case HttpStatusCode.Moved:
+                case HttpStatusCode.Found:
+                case HttpStatusCode.SeeOther:
+                case HttpStatusCode.TemporaryRedirect:
+                    needRedirect = true;
+                    break;
+
+                case HttpStatusCode.MultipleChoices:
+                    needRedirect = response.Headers.Location != null; // Don't redirect if no Location specified
+                    break;
+            }
+
+            return needRedirect;
+        }
+
+        private static bool RequestRequiresForceGet(HttpStatusCode statusCode, HttpMethod requestMethod)
+        {
+            if (statusCode == HttpStatusCode.Moved ||
+                statusCode == HttpStatusCode.Found ||
+                statusCode == HttpStatusCode.SeeOther ||
+                statusCode == HttpStatusCode.MultipleChoices)
+            {
+                return requestMethod == HttpMethod.Post;
+            }
+
+            return false;
+        }
+
+        private async Task<HttpResponseMessage> SendRequestAsync(HttpRequestMessage request, bool useCredentialCache, CancellationToken cancellationToken)
+        {
+            if (_credentials != null && _preAuthenticate)
+            {
+                // Try using previous digest response WWWAuthenticate header
+                if (_digestResponse != null)
+                {
+                    await AuthenticationHelper.TrySetDigestAuthToken(request, _credentials, _digestResponse, HttpKnownHeaderNames.Authorization).ConfigureAwait(false);
+                }
+                else
+                {
+                    AuthenticationHelper.TrySetBasicAuthToken(request, _credentials);
+                }
+            }
+
+            HttpResponseMessage response = await _innerHandler.SendAsync(request, cancellationToken).ConfigureAwait(false);
+            ICredentials currentCredential = useCredentialCache ? _credentials as CredentialCache : _credentials;
+
+            if (currentCredential != null && !_preAuthenticate && response.StatusCode == HttpStatusCode.Unauthorized)
+            {
+                HttpHeaderValueCollection<AuthenticationHeaderValue> authenticateValues = response.Headers.WwwAuthenticate;
+
+                foreach (AuthenticationHeaderValue h in authenticateValues)
+                {
+                    // We only support Basic and digest auth, ignore others
+                    if (h.Scheme == AuthenticationHelper.Basic)
+                    {
+                        if (AuthenticationHelper.TrySetBasicAuthToken(request, currentCredential))
+                        {
+                            response.Dispose();
+                            response = await _innerHandler.SendAsync(request, cancellationToken).ConfigureAwait(false);
+                            break;
+                        }
+                    }
+                    else if (h.Scheme == AuthenticationHelper.Digest)
```

Comment:
Digest should be preferred over Basic. Rather than a foreach over the header, how about a Contains?

Reply:
Is there a reference to an rfc guideline on this? As far as I remember, the first one in the string is preferred, and this code will pick the first one it sees while iterating.

Reply:
>Is there a reference to an rfc guideline on this? As far as I remember, the first one in the string is preferred, and this code will pick the first one it sees while iterating.

We never use the "first".  We pick from the list based on strength.  And Digest is stronger than Basic.

Same rules apply for NTLM, Negotiate, etc.  See WinHttpHandler for the ranking of schemes in terms of strength.

Reply:
Hmm ok, I was not aware that we did this in the other implementations.